File: python/08_071.py
class ClothingItem:
  def __init__(self, type, popularity):
     self.type = type
     self.popularity = popularity

class FashionStack:
  def __init__(self):
     self.mainStack = []
     self.tempStack = []
  
  def addItem(self, type, popularity):
     self.mainStack.append(ClothingItem(type, popularity))
  
  def removeTopItem(self):
     if not self.mainStack:
        return None
     return self.mainStack.pop()
  
  def findAndRemoveType(self, targetType):
     if not self.mainStack:
        return None
     
     mostPopular = None
     mostPopularIndex = -1
     
     # Find most popular item of the target type
     for i in range(len(self.mainStack) - 1, -1, -1):
        if self.mainStack[i].type == targetType:
           if (mostPopular is None or 
               self.mainStack[i].popularity >                
                      mostPopular.popularity):
              mostPopular = self.mainStack[i]
              mostPopularIndex = i
     
     if mostPopularIndex == -1:
        return None
     
     # Move items to temp stack until reaching target item
     for _ in range(len(self.mainStack) - mostPopularIndex - 1):
        self.tempStack.append(self.mainStack.pop())
     
     # Remove the target item
     result = self.mainStack.pop()
     
     # Restore items from temp stack
     while self.tempStack:
        self.mainStack.append(self.tempStack.pop())
     
     return result
  
  def reorganizeByPopularity(self):
     if not self.mainStack:
        return
     
     # Move all items to temp stack
     while self.mainStack:
        self.tempStack.append(self.mainStack.pop())
     
     # Sort items by popularity and restore to main stack
     sortedItems = sorted(
        self.tempStack, 
        key=lambda x: x.popularity, 
        reverse=True
     )
     self.tempStack = []
     
     for item in sortedItems:
        self.mainStack.append(item)

File: java/08_071.java
import java.util.*;

class ClothingItem {
   String type;
   int popularity;
   
   public ClothingItem(String type, int popularity) {
      this.type = type;
      this.popularity = popularity;
   }
}

public class FashionStack {
   private Stack<ClothingItem> mainStack;
   private Stack<ClothingItem> tempStack;
   
   public FashionStack() {
      mainStack = new Stack<>();
      tempStack = new Stack<>();
   }
   
   public void addItem(String type, int popularity) {
      mainStack.push(new ClothingItem(type, popularity));
   }
   
   public ClothingItem removeTopItem() {
      if (mainStack.isEmpty()) {
         return null;
      }
      return mainStack.pop();
   }
   
   public ClothingItem findAndRemoveType(String targetType) {
      if (mainStack.isEmpty()) {
         return null;
      }
      
      // Find most popular target type item
      ClothingItem mostPopular = null;
      int mostPopularIndex = -1;
      
      for (int i = mainStack.size() - 1; i >= 0; i--) {
         ClothingItem item = mainStack.get(i);
         if (item.type.equals(targetType)) {
            if (mostPopular == null || 
                  item.popularity > mostPopular.popularity) {
               mostPopular = item;
               mostPopularIndex = i;
            }
         }
      }
      
      if (mostPopularIndex == -1) {
         return null;
      }
      
      // Move items before target to temp stack
      int itemsToMove = mainStack.size() - mostPopularIndex - 1;
      for (int i = 0; i < itemsToMove; i++) {
         tempStack.push(mainStack.pop());
      }
      
      // Remove target item
      ClothingItem result = mainStack.pop();
      
      // Restore items from temp stack
      while (!tempStack.isEmpty()) {
         mainStack.push(tempStack.pop());
      }
      
      return result;
   }
   
   public void reorganizeByPopularity() {
      if (mainStack.isEmpty()) {
         return;
      }
      
      // Move all items to temp stack
      while (!mainStack.isEmpty()) {
         tempStack.push(mainStack.pop());
      }
      
      // Sort items by popularity
      List<ClothingItem> sortedItems = 
            new ArrayList<>(tempStack);
      sortedItems.sort((a, b) -> 
            Integer.compare(b.popularity, a.popularity));
      tempStack.clear();
      
      // Restore sorted items to main stack
      for (ClothingItem item : sortedItems) {
         mainStack.push(item);
      }
   }
}